community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -312,7 +316,7 @@ public void markForTermination()
             if ( stillSameTransaction && canBeTerminated() )
             {
                 failure = true;
-                terminated = true;
+                terminationReason = reason;
```

Comment:
I'm suspecting the `terminationReleaseLock` should cover the entire body of this method (so this includes the `canBeTerminated` check), and the entire body of the `close()` method.

Calls to `release()` can otherwise sneak in between this `canBeTerminated` check, and our first load of the `reuseCount` field, allowing us to terminate a transaction that has been released back to the pool, and possibly lend back out again.